### PR TITLE
Fix test_harness, increase speed

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

#231

## Proposed Changes

- Fix test harness
- Add caching for crosslink commitees.
- Fix a few bugs in our state transition logic.

## Additional Info

I've added a `CachedBeaconState` object which, whilst the concept is good, the implementation sucks because it fully duplicates the `BeaconState`. I've run out of time to do something which uses a reference.

It does significantly improve the speed of `test_harness` tests though, so I'm in favor of leaving it in for now.

I went on a journey to build a fully functional replacement to this, but it turns out `get_shuffling` is designed to to provide optimal lookups in the direction of `committee_index -> validator_index`, not `validator_index -> committee_index`.

TL;DR, this implementation sucks but it sucks _less_ than what we were doing before.

My recommendation is to merge it in now, then fix it later.

### CODE REVIEW

File: eth2/types/src/spec/few_validators.rs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+use crate::{ChainSpec, Slot};
```
Comment (AgeManning) [inline eth2/types/src/spec/few_validators.rs]: For consistency, should we put this inside `mod.rs`, or split everything out into `chain_spec.rs`. Either way combine the struct with the implementations. 
I guess this goes with `foundation.rs` as well. 
I see the benefit of separating as they show distinct logic, but like we did with `beacon_chain.rs` I think combining the impls with the structs throughout the project is good. 
Alternatively, having a directory structure like:
```
chain_spec.rs
chain_spec/
   foundation.rs
   few_validators.rs
```
could be useful for structs with large impl's.

Comment (paulhauner) [reply eth2/types/src/spec/few_validators.rs]: Yeah good call. I think keeping them in the same file here works -- `few_validators()` uses logic from `foundation()` so it's reasonable they live together.

This is a relic from before we standardized.

**Comment by paulhauner** [comment]
I moved the `ChainSpec` struct and all impls into `chain_spec.rs`.

Issue for `CachedBeaconState` references is at #237.

**Comment by AgeManning** [comment]
Nice
